synapse/federation/federation_client.py

```diff
@@ -1364,13 +1364,59 @@ async def send_request(
 
             return room, children, inaccessible_children
 
-        # TODO Fallback to the old federation API and translate the results.
-        return await self._try_destination_list(
-            "fetch room hierarchy",
-            destinations,
-            send_request,
-            failover_on_unknown_endpoint=True,
-        )
+        try:
+            return await self._try_destination_list(
+                "fetch room hierarchy",
+                destinations,
+                send_request,
+                failover_on_unknown_endpoint=True,
+            )
+        except SynapseError as e:
+            # Fallback to the old federation API and translate the results if
+            # no servers implement the new API.
+            #
+            # The algorithm below is a bit inefficient as it only attempts to
+            # get information for the requested room, but the legacy API may
+            # return additional layers.
+            if e.code == 502:
+                legacy_result = await self.get_space_summary(
+                    destinations,
+                    room_id,
+                    suggested_only,
+                    max_rooms_per_space=None,
+                    exclude_rooms=[],
+                )
+
+                # Find the requested room in the response (and remove it).
+                for _i, room in enumerate(legacy_result.rooms):
+                    if room.get("room_id") == room_id:
+                        break
+                else:
+                    # The requested room was not returned, nothing we can do.
+                    raise
+                requested_room = legacy_result.rooms.pop(_i)
+
+                # Find any children events of the requested room.
+                children_events = []
+                children_room_ids = set()
+                for event in legacy_result.events:
+                    if event.room_id == room_id:
+                        children_events.append(event.data)
+                        children_room_ids.add(event.state_key)
+                # And add them under the requested room.
+                requested_room["children_state"] = children_events
+
+                # Find the children rooms.
+                children = []
+                for room in legacy_result.rooms:
+                    if room.get("room_id") in children_room_ids:
+                        children.append(room)
+
+                # It isn't clear from the response whether some of the rooms are
+                # not accessible.
+                return requested_room, children, ()
```

Comment:
Will this produce odd results on the client-side where one may believe they can access a room, but will find they are unable to?

Mostly a question out of curiosity, as I don't know any way that the situation could be improved (and this is a legacy API so meh).

Reply:
It shouldn't, this is before any filtering is done on whether a room is accessible or not, it pretty much is just about changing API shapes from the old `/spaces` endpoint to the new `/hierarchy`. Part of that is that the `/spaces` endpoint returned multiple layers deep (over federation) while the `/hierarchy` one does *not*. So we throw away a bunch of data.

The overall algorithm is:

1. Find the requested room (which is returned in `rooms` for `/spaces`, but as a separate field for `/hierarchy`)
2. Attach the children events to it (which is returned in a list including children for other rooms in `/spaces` and as part of the `room` object for `/hierarchy`).
3. Get the summaries for any of those children room (which are returned as `rooms` for `/spaces`, but those will include additional layers).

Reply:
Aha, I see. Thanks for explaining it so clearly.